components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,245 @@
+                    largestkey: live_files.get_largestkey(i as i32).to_owned(),
+                    overlap_region_ids: Vec::new(),
+                    start_time: Instant::now(),
+                };
+
+                // only support recovering data key for now.
+                if !keys::validate_data_key(&f.smallestkey)
+                    || (!keys::validate_data_key(&f.largestkey)
+                        && f.largestkey.as_slice() != keys::DATA_MAX_KEY)
+                {
+                    set_panic_mark_and_panic(
+                        path,
+                        &format!(
+                            "key range mismatch, small key:{:?}, large key:{:?}",
+                            &f.smallestkey, &f.largestkey
+                        ),
+                    );
+                }
+
+                // lock the store_meta and check if the range exist.
+                if !self.check_overlap_damaged_regions(f) {
+                    set_panic_mark_and_panic(path, "Can't find overlap regions");
+                }
+
+                return;
+            }
+        }
+    }
+
+    // `sst_path` has been processed and is still in a damaged state.
+    fn exist_scheduling_regions(&self, sst_path: &str) -> bool {
+        self.damaged_files.iter().any(|f| f.name == sst_path)
+    }
+
+    // Periodically check for recorded damaged files。
+    //
+    // Acquire meta lock.
+    fn check_damaged_files(&mut self) {
+        for index in 0..self.damaged_files.len() {
+            if self.damaged_files[index].start_time.elapsed() > self.max_damage_duration {
+                let f = self.damaged_files.remove(index);
+                let fname = f.name.clone();
+                // file may be put into `damaged_files` again here, but when this happens,
+                // panic will be triggered, so it doesn’t matter.
+                if self.check_overlap_damaged_regions(f) {
+                    set_panic_mark_and_panic(
+                        &fname,
+                        &format!("recovery job exceeded {:?}", self.max_damage_duration),
+                    );
+                }
+            }
+        }
+    }
+
+    // Check whether the StoreMeta contains the region range, if it contains,
+    // recorded fault region ids to report to PD and add file info into `damaged_files`.
+    //
+    // Acquire meta lock.
+    fn check_overlap_damaged_regions(&mut self, mut file: FileInfo) -> bool {
+        let mut meta = match self.store_meta.lock() {
+            Ok(meta) => meta,
+            Err(e) => {
+                set_panic_mark_and_panic(&file.name, &format!("{}", e));
+                return false;
+            }
+        };
+
+        let mut overlap = false;
+
+        // Find overlapping region ids.
+        // Condition:
+        // end_key > file.smallestkey
+        // start_key <= file.largestkey
+        let mut ids = vec![];
+        for (_, id) in meta
+            .region_ranges
```

Comment:
Does the keys of `region_ranges` have the data prefix?

Reply:
yes, for example, we can found `meta.region_ranges.insert(enc_end_key(region), region_id);` and `enc_end_key` is implemented here
```
pub fn enc_end_key(region: &Region) -> Vec<u8> {
    // only initialized region's end_key can be encoded, otherwise there must be bugs
    // somewhere.
    assert!(!region.get_peers().is_empty());
    data_end_key(region.get_end_key())
}
```